enhancements/monitoring/replace-prometheus-adapter-with-metrics-server.md

```diff
@@ -0,0 +1,257 @@
+
+### Risks and Mitigations
+
+When considering equal scrape intervals, both metrics-server and prometheus-adapter have a comparable impact. This similarity arises from the fact that both applications perform a kubelet scrape once per instance within the designated scrape interval.
+
+We compared cpu and memory usage of metric-server vs prometheus-adapter on cluster-bot's cluster in [MON-3210](https://issues.redhat.com/browse/MON-3210) which didn't see significant difference when prometheus-adapter was replaced with metrics-server.
+
+We need to do performance testing on a larger OpenShift cluster (more than 100 nodes) and more than 70 pods on per node to measure the performance of Metrics Server. This is based on scalability table mentioned in [metrics-server](https://github.com/kubernetes-sigs/metrics-server#scaling). [MON-3328](https://issues.redhat.com/browse/MON-3328) is created for doing this.
+
+### Drawbacks
+
+N/A
+
+
+## Design Details
+
+### Open Questions [optional]
+
+### Test Plan
+
+The proposed solution will be tested with two suites:
+
+* Unit tests - running independently of each other and without side effects
+* End to end tests - in the same repository as the cluster-monitoring-operator. These tests will be executed against an actual Kubernetes cluster and are intended to verify end-to-end functionality of the entire system.
+
+E2E Test will test following:
+* Enables metrics-server
+* Verifies that the resource metrics API is working
+* Disables metrics-server
+* Verifies that the resource metrics API is working
+* During all the test, it will ensure that there's no disruption of the resource metrics API.
+
+### Graduation Criteria
+
+The main rationale behind adopting a phased approach is to prevent customers from encountering problems with the metrics-server without a contingency plan in place. This approach ensures a mechanism for reverting to the previous/legacy deployment if needed.
+
+* 4.15 TechPreview
+  * Notice in the release notes that prometheus-adapter is deprecated (should be repeated on every future release).
+* 4.16 GA as opt-in
+* 4.17 GA as opt-in, warn about prometheus-adapter being deprecated if `prometheus-adapter` config isn’t empty and `metrics-server` isn’t already enabled. CMO should report `Upgradeable=false` if the prometheus-adapter config isn't empty.
+* 4.18 GA, enabled by default
+* 4.19 => Remove `prometheus-adapter` from CMO
+
+
+#### Dev Preview -> Tech Preview
+
+N/A
+
+#### Tech Preview -> GA
+
+More testing (upgrade, downgrade, scale).
+* Sufficient time for feedback including signals from telemetry about the customer adoption
+  * We will be adding a Gauge Metric named `cluster_monitoring_operator_metrics_server_enabled` in [cmo](https://github.com/openshift/cluster-monitoring-operator/blob/master/pkg/metrics/metrics.go) to measure this.
+* Counter-measures to avoid service degradation of the Kubelet.
```

Comment:
I wonder if this really makes sense to do.

Scraping kubelet is only fragile on SNO and since Prometheus will not be scraping kubelet there anymore, having metrics-server would have the same impact as the current setup

Reply:
We'll still have Prometheus scraping kubelet's metrics from /metrics/cadvisor (cAdvisor) and /metrics for generic monitoring & alerting.

Reply:
Oh I see, reading the enhancement I somehow thought that only metrics-server would be kept on SNO